### PR TITLE
[Logstash] Add plugin specific drilldown dashboards

### PR DESCRIPTION
**Pull request by robbavey**
## Proposed commit message

This PR adds additional drilldown dashboards on input, filter and output plugin in order to to add contextual information about where a plugin was defined. This is important, as it allows users to quickly find out which of many plugins
of the same type are being referenced in the dashboards, without needing to rewrite long and complex pipeline configurations to add identifiers to each of their plugin definitions.

To do this, the CEL code is augmented to retrieve pipeline information from the /pipelines/#{pipeline_name} endpoint, which provides information about each of the pipeline graphs.

Additionally, this PR adds CEL code to pull metrics data from the pipeline graph, and make this available to plugin specific drilldowns, with dashboards for `logstash-filter-dissect`, `logstash-filter-grok` and `logstash-output-elasticsearch` provided here.

Additionally, this commit includes the necessary field mappings for the `dissect` and `grok` filters and `elasticsearch` output, and drilldown dashboards  for `dissect`, `grok` and the `elasticsearch` output and the necessary links from the single pipeline view to enable access to these dashboards.

This also includes improvements to dashboard organization and consistency, ensuring
that dashboards are all organized in the same way with "events received" always being
shown before "events emitted" and organized vertically for improved comparisons.
There are also some updates made for consistent labelling of visualizations.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Screenshots

Drilldown dashboards are created for `input`, `filter`, and `output` plugin types, with links added from the plugin tables from the `Single Pipeline View` dashboard.

Generic filter plugin drilldown 
Screenshot 2023-10-20 at 2 42 25 PM

Elasticsearch Output Dashboard
Screenshot 2023-10-20 at 4 41 18 PM

Grok Filter Dashboard
Screenshot 2023-10-20 at 4 41 43 PM

### CODE REVIEW

**Comment by robbavey** [comment]
@efd6 Sorry to bother you again, but there is some more CEL-based work for Logstash dashboards, if you wouldn't mind taking a look

@klacabane, @miltonhultgren and some more logstash dashboard content, too

File: packages/logstash/data_stream/pipeline/agent/stream/cel.yml.hbs
Comment (efd6) [inline packages/logstash/data_stream/pipeline/agent/stream/cel.yml.hbs:1]: Style; prefer vertical over horizontal.
```
program: |
  get(state.url).as(resp, bytes(resp.Body).decode_json().as(body,
    body.pipelines.map(pipeline_name, pipeline_name != ".monitoring-logstash", {
      "name": pipeline_name,
      "elasticsearch.cluster.id": has(body.pipelines[pipeline_name].vertices) ?
        body.pipelines[pipeline_name].vertices.map(vertex, has(vertex.cluster_uuid), vertex.cluster_uuid))
      :
        [],
      "host":{
        "name":body.name,
        "address":body.http_address,
      },
      "total":{
        "flow":body.pipelines[pipeline_name].flow,
        "time":{
          "queue_push_duration": {
            "ms": has(body.pipelines[pipeline_name].events.queue_push_duration_in_millis) ?
              body.pipelines[pipeline_name].events.queue_push_duration_in_millis
            :
              [],
          },
          "duration":{
            "ms": has(body.pipelines[pipeline_name].events.duration_in_millis) ?
              body.pipelines[pipeline_name].events.duration_in_millis
            :
              [],
          },
        },
        "reloads":{
          "successes":body.pipelines[pipeline_name].reloads.successes,
          "failures":body.pipelines[pipeline_name].reloads.failures
        },
        "events":{
          "out": has(body.pipelines[pipeline_name].events.out) ?
            body.pipelines[pipeline_name].events.out
          :
            [],
          "in": has(body.pipelines[pipeline_name].events.out) ? // This deliberately uses 'out' as `has` does not accept `in`
            body.pipelines[pipeline_name].events['in']
          :
            [],
          "filtered": has(body.pipelines[pipeline_name].events.filtered) ?
            body.pipelines[pipeline_name].events.filtered
          :
            [],
        },
        "queues":{
          "type": has(body.pipelines[pipeline_name].queue.type) ?
            body.pipelines[pipeline_name].queue.type
          :
            [],
          "events": has(body.pipelines[pipeline_name].queue.events_count) ?
            body.pipelines[pipeline_name].queue.events_count
          :
            [],
          "current_size":{
            "bytes": has(body.pipelines[pipeline_name].queue.queue_size_in_bytes) ?
              body.pipelines[pipeline_name].queue.queue_size_in_bytes
            :
              [],
          },
          "max_size":{
            "bytes": has(body.pipelines[pipeline_name].queue.max_queue_size_in_bytes) ?
              body.pipelines[pipeline_name].queue.max_queue_size_in_bytes
            :
              [],
          }
        }
      }
    }))).as(pipelines, {
      "events": pipelines.map(pipeline, {
        "logstash": {"pipeline":pipeline}
      })
    })
```

File: packages/logstash/data_stream/pipeline/agent/stream/cel.yml.hbs [hunk 1]
```diff
@@ -23,7 +23,7 @@ program: |
   .decode_json().as(body,
     body.pipelines.map(pipeline_name, pipeline_name != ".monitoring-logstash", {"name":pipeline_name}
       .with({
-        "elasticsearch.cluster.id":((body.pipelines[pipeline_name].vertices).as(vertices, vertices.map(each, has(each.cluster_uuid), each.cluster_uuid))),
+        "elasticsearch.cluster.id":has(body.pipelines[pipeline_name].vertices) ? body.pipelines[pipeline_name].vertices.as(vertices, vertices.map(each, has(each.cluster_uuid), each.cluster_uuid)) : [],
```
Comment (efd6) [inline packages/logstash/data_stream/pipeline/agent/stream/cel.yml.hbs:26]: This doesn't need an `as`. In general, `as` should be used in two situations only.
1. Where the value is needed more than once in the subexpression.
2. Where there is a significant clarity improvement by given the value a label.

The 2. comes down to style, but here I think you get all the clarity you need by making the iterator value take the name `vertex`.

File: packages/logstash/data_stream/pipeline/agent/stream/cel.yml.hbs [hunk 2]
```diff
@@ -32,29 +32,30 @@ program: |
...
           "reloads":{
             "successes":body.pipelines[pipeline_name].reloads.successes,
             "failures":body.pipelines[pipeline_name].reloads.failures
           },
           "events":{
-            "out":body.pipelines[pipeline_name].events.out,
-            "in":body.pipelines[pipeline_name].events["in"],
-            "filtered":body.pipelines[pipeline_name].events.filtered,
+            "out":has(body.pipelines[pipeline_name].events.out) ? body.pipelines[pipeline_name].events.out : [],
+            // This deliberately uses 'out' as `has` does not accept `in`
```
Comment (efd6) [inline packages/logstash/data_stream/pipeline/agent/stream/cel.yml.hbs:47]: Is this a language thing? What happens?

Have you tried `("in" in body.pipelines[pipeline_name].events)`? I'm not sure it's any clearer, so if the comment is good enough for you, it's reasonable.

File: packages/logstash/data_stream/plugins/agent/stream/cel.yml.hbs
Comment (efd6) [inline packages/logstash/data_stream/plugins/agent/stream/cel.yml.hbs:1]: Style and some efficiency (hoisting filters and flatten operations).
```
program: |
  get(state.url + "/stats?graph=true&vertices=true").as(resp, bytes(resp.Body).decode_json().as(body,
    body.pipelines.map(pipeline_name, pipeline_name != ".monitoring-logstash", body.pipelines[pipeline_name].with({
      "name":pipeline_name
      "pipeline_source_map":
        get(state.url + "/pipelines/" + pipeline_name + "?graph=true&vertices=true").as(resp,
          bytes(resp.Body).decode_json().as(pipes,
            has(pipes.pipeline) ?
              pipes.pipelines.map(pipeline_name,
                has(pipes.pipelines) && pipes.pipelines != null && pipes.pipelines[pipeline_name].graph.graph.vertices != null,
                pipes.pipelines[pipeline_name].graph.graph.vertices.map(vertex, vertex.type == "plugin", {
                  "plugin_id": vertex.id,
                  "source": vertex.meta.source,
                })
              ).drop("graph").flatten()
            :
              []
        )
      ),
      "es_cluster_id": has(body.pipelines[pipeline_name].vertices) ?
        body.pipelines[pipeline_name].vertices.map(vertex, has(vertex.cluster_uuid), vertex.cluster_uuid)
      :
        [],
      "es_cluster_id_map": has(body.pipelines[pipeline_name].vertices) ?
        body.pipelines[pipeline_name].vertices.map(vertex, has(vertex.cluster_uuid), {
          "plugin_id": vertex.id,
          "cluster_id": vertex.cluster_uuid,
        })
      :
        [],
      "counter_map": has(body.pipelines[pipeline_name].vertices) ?
        body.pipelines[pipeline_name].vertices.map(vertex, has(vertex.long_counters), vertex.long_counters.map(counter, {
          "plugin_id": vertex.id,
          "name": counter.name,
          "value": counter.value
        }))
      :
        [],
      "outputs": body.pipelines[pipeline_name].plugins.outputs,
      "inputs": body.pipelines[pipeline_name].plugins.inputs,
      "filters": body.pipelines[pipeline_name].plugins.filters,
      "codecs": body.pipelines[pipeline_name].plugins.codecs,
      "host":{
        "name": body.name,
        "address": body.http_address,
      }
    })))).as(events, events.map(event, {
      "inputs": event.inputs.map(input, has(event.hash), {
        "name": event.name,
        "id": event.hash,
        "host": event.host,
        "elasticsearch.cluster.id": event.es_cluster_id,
        "plugin": {
          "type": "input",
          "input": {
            "source": event.pipeline_source_map.map(tuple, tuple.plugin_id == input.id, tuple.source)[0],
            "elasticsearch.cluster.id": event.es_cluster_id_map.map(tuple, tuple.plugin_id == input.id, tuple.cluster_id),
            "metrics": {
              input.name: event.counter_map.flatten().filter(tuple, tuple.plugin_id == input.id).as(counter_map, zip(
                counter_map.map(tuple, tuple.name),
                counter_map.map(tuple, tuple.value),
              ))
             },
            "name": input.name,
            "id": input.id,
            "flow": has(input.flow) ?
              input.flow
            :
              {},
            "events": {
              "out": input.events.out,
            },
            "time": {
              "queue_push_duration": {
                "ms": input.events.queue_push_duration_in_millis
              }
            }
          }
        }
      }.drop_empty()),
      "codecs": event.codecs.map(codec, has(event.hash), {
        "name": event.name,
        "id": event.hash,
        "host": event.host,
        "elasticsearch.cluster.id": event.es_cluster_id,
        "plugin": {
          "type": "codec",
          "codec": {
          "id":codec.id,
          "name":codec.name,
            "flow": has(codec.flow) ? codec.flow : {},
            "decode":{
              "duration":{
                "ms":codec.decode.duration_in_millis
              },
              "in":codec.decode.writes_in,
              "out":codec.decode.out,
            },
            "encode":{
              "in":codec.encode.writes_in,
              "duration":{
                "ms":codec.encode.duration_in_millis
              }
            }
          }
        }
      }.drop_empty()),
      "filters": event.filters.map(filter, has(event.hash), {
        "name": event.name,
        "id": event.hash,
        "host": event.host,
        "elasticsearch.cluster.id": event.es_cluster_id,
        "plugin": {
          "type": "filter",
          "filter": {
            "source": event.pipeline_source_map.map(tuple, tuple.plugin_id == filter.id, tuple.source).flatten()[0],
            "id": filter.id,
            "name": filter.name,
            "elasticsearch.cluster.id": event.es_cluster_id_map.map(tuple, tuple.plugin_id == filter.id, tuple.cluster_id),
            "metrics": {
              filter.name: event.counter_map.flatten().filter(tuple, tuple.plugin_id == input.id).as(counter_map, zip(
                counter_map.map(tuple, tuple.name),
                counter_map.map(tuple, tuple.value),
              ))
            },
            "flow": has(filter.flow) ?
              filter.flow
            :
              {},
            "events": {
              "in": filter.events['in'],
              "out": filter.events.out,
            },
            "time": {
              "duration": {
                "ms": filter.events.duration_in_millis
              }
            }
          }
        }
      }.drop_empty()),
      "outputs": event.outputs.map(output, has(event.hash), {
        "name": event.name,
        "id": event.hash,
        "host": event.host,
        "elasticsearch.cluster.id": event.es_cluster_id,
        "plugin": {
          "type": "output",
          "output": {
            "id": output.id,
            "name": output.name,
            "source": event.pipeline_source_map.map(tuple, tuple.plugin_id == output.id, tuple.source).flatten()[0],
            "elasticsearch.cluster.id": event.es_cluster_id_map.map(tuple, tuple.plugin_id == output.id, tuple.cluster_id),
            "metrics": {
              output.name: event.counter_map.flatten().filter(tuple, tuple.plugin_id == input.id).as(counter_map, zip(
                counter_map.map(tuple, tuple.name),
                counter_map.map(tuple, tuple.value),
              ))
            },
            "flow": has(output.flow) ?
              output.flow
            :
              {},
            "events":{
              "in":output.events['in'],
              "out":output.events.out,
            },
            "time":{
              "duration":{
                "ms":output.events.duration_in_millis
              }
            }
          }
        }
      }.drop_empty())
    }).collate(["filters", "outputs", "inputs", "codecs"])).as(plugins, {
      "events": plugins.map(plugin, {
        "logstash":{"pipeline":plugin}
      })
    })
```

File: packages/logstash/data_stream/plugins/agent/stream/cel.yml.hbs [hunk 3]
```diff
@@ -45,7 +59,10 @@ program: |
         "plugin":{
           "type":"input",
           "input":{
+            "source":event.pipeline_source_map.map(tuple, (tuple.plugin_id == input.id), tuple.source)[0],
```
Comment (efd6) [inline packages/logstash/data_stream/plugins/agent/stream/cel.yml.hbs:62]: Is this safe? (also below)

File: packages/logstash/data_stream/plugins/agent/stream/cel.yml.hbs [hunk 4]
```diff
@@ -45,7 +59,10 @@ program: |
...
             "elasticsearch.cluster.id":event.es_cluster_id_map.map(tuple, (tuple.plugin_id == input.id), tuple.cluster_id),
+            "metrics":{input.name:zip(event['counter_map'].flatten().map(tuple, (tuple.plugin_id == input.id), tuple.name),
```
Comment (efd6) [inline packages/logstash/data_stream/plugins/agent/stream/cel.yml.hbs:64]: Each `flatten` is a Go map allocation, so do this only once. Also reduce `map` use counts.

**Review by efd6 — COMMENTED** [review]
Suggestions are untested.

**Review by klacabane — APPROVED** [review]
LGTM!

File: packages/logstash/data_stream/plugins/agent/stream/cel.yml.hbs (hunk 3 above)
Comment (robbavey) [reply packages/logstash/data_stream/plugins/agent/stream/cel.yml.hbs:62]: I believe so - I have tested with all the permutations of the source, including none available at all

File: packages/logstash/data_stream/pipeline/agent/stream/cel.yml.hbs (hunk 2 above)
Comment (robbavey) [reply packages/logstash/data_stream/pipeline/agent/stream/cel.yml.hbs:47]: Depending on whether I use `events.in` or `events['in']`, I get 
```
failed program instantiation: failed compilation: ERROR: <input>:35:58: Syntax error: no viable alternative at input '.in'
 |           "in": has(body.pipelines[pipeline_name].events.in) 
 ```
 or
 ```
 failed program instantiation: failed compilation: ERROR: <input>:35:20: invalid argument to has() macro
 |           "in": has(body.pipelines[pipeline_name].events['in'])
 ```

However, your suggestion of using `"in": ("in" in body.pipelines[pipeline_name].events) `, works perfectly. I'll switch

Comment (robbavey) [reply packages/logstash/data_stream/pipeline/agent/stream/cel.yml.hbs]: I ended up reverting that change - I with further testing, I was seeing 

```
failed eval: no such overload
```

when using `"in": ("in" in body.pipelines[pipeline_name].events)`

**Comment by robbavey** [comment]
Ready for another round I think @efd6

File: packages/logstash/data_stream/pipeline/agent/stream/cel.yml.hbs [hunk 5]
```diff
@@ -18,48 +18,76 @@ redact:
...
+    }))).as(pipelines, {
+      "events": pipelines.map(pipeline, {
+        "logstash": {"pipeline":pipeline}
       })
-    )
-    ))
-    .as(pipelines, {
-      "events":pipelines.map(pipeline, {"logstash":{"pipeline":pipeline}})})
\ No newline at end of file
+    })
```
Comment (efd6) [inline packages/logstash/data_stream/pipeline/agent/stream/cel.yml.hbs:93]: Nit: final new line.

File: packages/logstash/data_stream/plugins/agent/stream/cel.yml.hbs [hunk 6]
```diff
@@ -140,7 +190,9 @@ program: |
...
-      }.drop_empty()
-      )
+      }.drop_empty())
     }).collate(["filters", "outputs", "inputs", "codecs"])).as(plugins, {
-      "events":plugins.map(plugin, {"logstash":{"pipeline":plugin}})})
\ No newline at end of file
+      "events": plugins.map(plugin, {
+        "logstash":{"pipeline":plugin}
+      })
+    })
```
Comment (efd6) [inline packages/logstash/data_stream/plugins/agent/stream/cel.yml.hbs:198]: Final new line.

File: packages/logstash/data_stream/plugins/agent/stream/cel.yml.hbs [hunk 7]
```diff
@@ -17,59 +17,93 @@ redact:
...
+          "plugin_id": vertex.id,
+          "name": counter.name,
+          "value": counter.value
+        }))
+      :
+        [],
+      "outputs": body.pipelines[pipeline_name].plugins.outputs,
+      "inputs": body.pipelines[pipeline_name].plugins.inputs,
+      "filters": body.pipelines[pipeline_name].plugins.filters,
+      "codecs": body.pipelines[pipeline_name].plugins.codecs,
```
Comment (efd6) [inline packages/logstash/data_stream/plugins/agent/stream/cel.yml.hbs:60]: There's an opportunity for an optimisation here, but it requires some gymnastics, so I would suggest leaving it until you see a need. There are four look-ups into `body.pipelines` that all get to the same field, so you could prepare that when you prepare `pipeline_name`, but then you need another `as` to get the two parts into an object to refer to (this is the gymnastics).

**Review by efd6 — APPROVED** [review]
CEL code LGTM

**Pull request merged**